Sources/PackageGraph/CompoundPackageContainerProvider.swift

```diff
@@ -0,0 +1,81 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Dispatch
+
+import PackageLoading
+import PackageModel
+import PackageRegistry
+import SourceControl
+
+import TSCBasic
+import TSCUtility
+
+public class CompoundPackageContainerProvider: PackageContainerProvider {
+    let manifestLoader: ManifestLoaderProtocol
+    let repositoryManager: RepositoryManager
+    let mirrors: DependencyMirrors
+
+    /// The tools version currently in use. Only the container versions less than and equal to this will be provided by
+    /// the container.
+    let currentToolsVersion: ToolsVersion
+
+    /// The tools version loader.
+    let toolsVersionLoader: ToolsVersionLoaderProtocol
+
+    let fileSystem: FileSystem
+
+    /// Queue for callbacks.
+    private let callbacksQueue = DispatchQueue(label: "org.swift.swiftpm.container-provider")
+
+    /// Create a repository-based package provider.
+    ///
+    /// - Parameters:
+    ///   - repositoryManager: The repository manager responsible for providing repositories.
+    ///   - manifestLoader: The manifest loader instance.
+    ///   - currentToolsVersion: The current tools version in use.
+    ///   - toolsVersionLoader: The tools version loader.
+    public init(
+        repositoryManager: RepositoryManager,
+        mirrors: DependencyMirrors = DependencyMirrors(),
+        manifestLoader: ManifestLoaderProtocol,
+        currentToolsVersion: ToolsVersion = ToolsVersion.currentToolsVersion,
+        toolsVersionLoader: ToolsVersionLoaderProtocol = ToolsVersionLoader(),
+        fileSystem: FileSystem = localFileSystem
+    ) {
+        self.repositoryManager = repositoryManager
+        self.mirrors = mirrors
+        self.manifestLoader = manifestLoader
+        self.currentToolsVersion = currentToolsVersion
+        self.toolsVersionLoader = toolsVersionLoader
+        self.fileSystem = fileSystem
+    }
+
+    public func getContainer(
+        for package: PackageReference,
+        skipUpdate: Bool,
+        completion: @escaping (Result<PackageContainer, Swift.Error>) -> Void
+    ) {
+        let provider: PackageContainerProvider
+
+        switch package {
+        case .root, .local:
+            provider = LocalPackageContainerProvider(mirrors: self.mirrors, manifestLoader: self.manifestLoader, currentToolsVersion: self.currentToolsVersion, toolsVersionLoader: self.toolsVersionLoader, fileSystem: self.fileSystem)
+        case .remote:
+            if let registryManager = (try? await { RegistryManager.discover(for: package, completion: $0) }) {
```

Comment:
this seems to be blocking on an async operation, and since `getContainer` is async we could avoid